src/filter/copy.c

```diff
@@ -0,0 +1,191 @@
+#include "core/assert.h"
+
+static core_log_t _log = LOG_T_INIT("filter.copy");
+static filter_copy_t _defaults = {
+    LOG_T_INIT_OBJ("filter.copy"),
+    0, 0,
+    0
+};
+
+core_log_t* filter_copy_log()
+{
+    return &_log;
+}
+
+void filter_copy_init(filter_copy_t* self)
+{
+    mlassert_self();
+
+    *self = _defaults;
+}
+
+void filter_copy_destroy(filter_copy_t* self)
+{
+    mlassert_self();
+}
+
+void filter_copy_set(filter_copy_t* self, int32_t obj_type)
+{
+    mlassert_self();
+
+    switch (obj_type) {
+    case CORE_OBJECT_NONE:
+        self->copy |= 0x1;
+        break;
+    case CORE_OBJECT_PCAP:
+        self->copy |= 0x2;
+        break;
+    case CORE_OBJECT_ETHER:
+        self->copy |= 0x4;
+        break;
+    case CORE_OBJECT_NULL:
+        self->copy |= 0x8;
+        break;
+    case CORE_OBJECT_LOOP:
+        self->copy |= 0x10;
+        break;
+    case CORE_OBJECT_LINUXSLL:
+        self->copy |= 0x20;
+        break;
+    case CORE_OBJECT_IEEE802:
+        self->copy |= 0x40;
+        break;
+    case CORE_OBJECT_GRE:
+        self->copy |= 0x80;
+        break;
+    case CORE_OBJECT_IP:
+        self->copy |= 0x100;
+        break;
+    case CORE_OBJECT_IP6:
+        self->copy |= 0x200;
+        break;
+    case CORE_OBJECT_ICMP:
+        self->copy |= 0x400;
+        break;
+    case CORE_OBJECT_ICMP6:
+        self->copy |= 0x800;
+        break;
+    case CORE_OBJECT_UDP:
+        self->copy |= 0x1000;
+        break;
+    case CORE_OBJECT_TCP:
+        self->copy |= 0x2000;
+        break;
+    case CORE_OBJECT_PAYLOAD:
+        self->copy |= 0x4000;
+        break;
+    case CORE_OBJECT_DNS:
+        self->copy |= 0x8000;
+        break;
+    default:
+        lfatal("unknown type %d", obj_type);
+    }
+}
+
+uint64_t filter_copy_get(filter_copy_t* self, int32_t obj_type)
+{
+    mlassert_self();
+
+    switch (obj_type) {
+    case CORE_OBJECT_NONE:
+        return self->copy & 0x1;
+    case CORE_OBJECT_PCAP:
+        return self->copy & 0x2;
+    case CORE_OBJECT_ETHER:
+        return self->copy & 0x4;
+    case CORE_OBJECT_NULL:
+        return self->copy & 0x8;
+    case CORE_OBJECT_LOOP:
+        return self->copy & 0x10;
+    case CORE_OBJECT_LINUXSLL:
+        return self->copy & 0x20;
+    case CORE_OBJECT_IEEE802:
+        return self->copy & 0x40;
+    case CORE_OBJECT_GRE:
+        return self->copy & 0x80;
+    case CORE_OBJECT_IP:
+        return self->copy & 0x100;
+    case CORE_OBJECT_IP6:
+        return self->copy & 0x200;
+    case CORE_OBJECT_ICMP:
+        return self->copy & 0x400;
+    case CORE_OBJECT_ICMP6:
+        return self->copy & 0x800;
+    case CORE_OBJECT_UDP:
+        return self->copy & 0x1000;
+    case CORE_OBJECT_TCP:
+        return self->copy & 0x2000;
+    case CORE_OBJECT_PAYLOAD:
+        return self->copy & 0x4000;
+    case CORE_OBJECT_DNS:
+        return self->copy & 0x8000;
+    default:
+        lfatal("unknown type %d", obj_type);
+    }
+}
```

Comment:
```
../../src/filter/copy.c:148: warning: control reaches end of non-void function
```

Reply:
I didn't see this warning with GCC 9.1.0, but it should be fixed now. On a related note, I get `gcc: warning: switch '-mmpx' is no longer supported`, but I haven't checked what it's supposed to do.

Reply:
No worry, build platforms are kinda strict on warnings. Also, I have `--enable-warn-all` for all projects, if you use that you might get the warning.